hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -3304,16 +3304,21 @@ private TableGroup consumeEntityJoin(SqmEntityJoin<?> sqmJoin, TableGroup lhsTab
 				sqmJoin.getNavigablePath(),
 				sqmJoin.getExplicitAlias(),
 				null,
-				() -> p -> predicate.set( combinePredicates( predicate.get(), p ) ),
+				() -> p -> needsTreat.setValue( true ),
 				this
-				);
+		);
 		getFromClauseIndex().register( sqmJoin, tableGroup );
 
+		if ( needsTreat.getValue() ) {
+			// Register new treat to apply the discriminator condition to the table reference itself, see #pruneTableGroupJoins
+			registerEntityNameUsage( tableGroup, EntityNameUse.TREAT, entityDescriptor.getEntityName() );
```

Comment:
Don't we need something similar for association joins as well? What happens if we do `from EntityA a right join a.association`?

Reply:
I'm not sure, but I can test this. I guess this could be a problem for associations that use a specific inheritance subtype, or even the root type but the query specifies a subtype-specific path thus implicitly treating it (though this second case should might already be covered by the new entity-name uses).

Reply:
I've added 3 more tests that do the same checks as the entity join ones but with an association. I also updated the logic to add a treat for an attribute join to an inheritance subtype entity to ensure we only retrieve records of that type.